### ISSUE TITLE
_canonical_mask throws warning when bool masks passed as input to TransformerDecoder

### ISSUE THREAD

**Issue by tomhosking** [pytorch/pytorch#96103]
### 🐛 Describe the bug

Similar to #95702, but for `TransformerDecoder` - passing `bool` masks results in a warning being thrown about mismatched mask types, as `_canonical_masks` is called multiple times.

```
import torch
import torch.nn as nn


def transformer_decoder(inputs, input_seq_len, memory):
    decoder_layer = nn.TransformerDecoderLayer(
        d_model=16,
        nhead=2,
        dim_feedforward=32,
        dropout=0.1,
        activation='relu',
        batch_first=True,
    )
    decoder_norm = nn.LayerNorm(16)
    decoder = nn.TransformerDecoder(
        decoder_layer, 2, decoder_norm
    )

    src_mask = torch.ones(inputs.shape[1], inputs.shape[1], dtype=torch.bool).triu_(diagonal=1)
    padding_mask = (torch.arange(inputs.shape[1])[None, :].cpu() >= input_seq_len[:, None])
    
    return decoder(inputs, 
                   memory,
        tgt_mask=src_mask,
       tgt_key_padding_mask=padding_mask,
        memory_key_padding_mask=padding_mask,
    )
    
inputs = torch.randn(2,3,16)
memory = torch.randn(2,3,16)
input_seq_len = torch.tensor([3,2])

transformer_decoder(inputs, input_seq_len, memory)
```

```
/mnt/ext/phd/research/pt2/lib/python3.9/site-packages/torch/nn/functional.py:5004: UserWarning: Support for mismatched key_padding_mask and attn_mask is deprecated. Use same type for both instead.
```

(@mikekgfb)

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230306+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.9.13 (main, Oct 13 2022, 21:15:33)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-67-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 525.85.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          8
On-line CPU(s) list:             0-7
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Core(TM) i7-9700 CPU @ 3.00GHz
CPU family:                      6
Model:                           158
Thread(s) per core:              1
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        13
CPU max MHz:                     4700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6000.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_
perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_
timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2
 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities
Virtualisation:                  VT-x
L1d cache:                       256 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        2 MiB (8 instances)
L3 cache:                        12 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-7
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT disabled
Vulnerability Retbleed:          Mitigation; Enhanced IBRS                                                                                                                   
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Mitigation; Microcode
Vulnerability Tsx async abort:   Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.0.0+b8b470bc59
[pip3] torch==2.1.0.dev20230306+cu118
[pip3] torchaudio==2.0.0.dev20230306+cu118
[pip3] torchvision==0.15.0.dev20230306+cu118
[conda] numpy                     1.21.5                   pypi_0    pypi
[conda] pytorch-triton            2.0.0+b8b470bc59          pypi_0    pypi
[conda] torch                     1.7.1                    pypi_0    pypi
[conda] torchseq                  3.0.0a0                   dev_0    <develop>

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by ngimel** [comment]
@mikekgfb this is probably fixed?

**Comment by tomhosking** [comment]
Still seeing this issue in the latest nightly build (2.1.0.dev20230307).

**Comment by mikekgfb** [comment]
Does the build. you're looking at contain this PR => https://github.com/pytorch/pytorch/pull/96009
This should also include your repro as a unit test, and it passes in CI test_padding_and_src_mask_bool in tests/test_transformer.py, so I'll appreciate if you can check this on your end?

**Comment by tomhosking** [comment]
I will double check when I can - but just to confirm that the issue I'm seeing now is with the Decoder, not Encoder, the previous issue I raised was for the Encoder only (which is now fixed).

**Comment by mikekgfb** [comment]
Can you please share a test case for your problem?  On the surface of it, decoder should not have been affected by the canonical mask issue, but it could have been triggered by a logic issue in the shared MHA block.

**Comment by tomhosking** [comment]
Running the code from #95702 (testing `TransformerEncoder`, fixed in #96009) with the latest nightly build works fine, no warning is thrown.

Running the snippet above (using `TransformerDecoder`) throws the warning from `_canonical_mask`, trace as below:

```
pt2_test.py:38: in <module>
    transformer_decoder(inputs, input_seq_len, memory)
pt2_test.py:25: in transformer_decoder
    return decoder(inputs,
pt2/lib/python3.9/site-packages/torch/nn/modules/module.py:1533: in _call_impl
    return forward_call(*args, **kwargs)
pt2/lib/python3.9/site-packages/torch/nn/modules/transformer.py:369: in forward
    output = mod(output, memory, tgt_mask=tgt_mask,
pt2/lib/python3.9/site-packages/torch/nn/modules/module.py:1533: in _call_impl
    return forward_call(*args, **kwargs)
pt2/lib/python3.9/site-packages/torch/nn/modules/transformer.py:716: in forward
    x = self.norm1(x + self._sa_block(x, tgt_mask, tgt_key_padding_mask, tgt_is_causal))
pt2/lib/python3.9/site-packages/torch/nn/modules/transformer.py:725: in _sa_block
    x = self.self_attn(x, x, x,
pt2/lib/python3.9/site-packages/torch/nn/modules/module.py:1533: in _call_impl
    return forward_call(*args, **kwargs)
pt2/lib/python3.9/site-packages/torch/nn/modules/activation.py:1210: in forward
    attn_output, attn_output_weights = F.multi_head_attention_forward(
pt2/lib/python3.9/site-packages/torch/nn/functional.py:5162: in multi_head_attention_forward
    key_padding_mask = _canonical_mask(
pt2/lib/python3.9/site-packages/torch/nn/functional.py:5004: in _canonical_mask
    warnings.warn(
E   UserWarning: Support for mismatched key_padding_mask and attn_mask is deprecated. Use same type for both instead.
```

**Comment by tomhosking** [comment]
Fixed for me with `2.1.0.dev20230313+cu118` - thanks!
